src/cargo/util/cache_lock.rs

```diff
@@ -0,0 +1,463 @@
+//!   collection runs.
+//! * [`CacheLockMode::MutateExclusive`] -- This is an exclusive lock acquired
+//!   whenever needing to modify existing source files (for example, with
+//!   cache garbage collection). This is acquired to make sure that no other
+//!   cargo is reading from the cache.
+//!
+//! Importantly, a `DownloadExclusive` lock does *not* interfere with a
+//! `Shared` lock. The download process generally does not modify source
+//! files, so other cargos should be able to safely proceed in reading source
+//! files[^1].
+//!
+//! This is implemented by two separate lock files, the "download" one and the
+//! "mutate" one. The `MutateExclusive` lock acquired both the "mutate" and
+//! "download" locks. The `Shared` lock acquires the "mutate" lock in share
+//! mode.
+//!
+//! An important rule is that `MutateExclusive` acquires the locks in the
+//! order "mutate" first and then the "download". That helps prevent
+//! deadlocks. It is not allowed for a cargo to first acquire a
+//! `DownloadExclusive` lock and then a `Shared` lock because that would open
+//! it up for deadlock.
+//!
+//! Another rule is that there should be only one [`CacheLocker`] per process
+//! to uphold the ordering rules. You could in theory have multiple if you
+//! could ensure that other threads would make progress and drop a lock, but
+//! cargo is not architected that way.
+//!
+//! It is safe to recursively acquire a lock as many times as you want.
+//!
+//! ## Interaction with older cargos
+//!
+//! Before version 1.74, cargo only acquired the `DownloadExclusive` lock when
+//! downloading and doing resolution. Newer cargos that acquire
+//! `MutateExclusive` should still correctly block when an old cargo is
+//! downloading (because it also acquires `DownloadExclusive`), but they do
+//! not properly coordinate when an old cargo is in the build phase (because
+//! it holds no locks). This isn't expected to be much of a problem because
+//! the intended use of mutating the cache is only to delete old contents
+//! which aren't currently being used. It is possible for there to be a
+//! conflict, particularly if the user manually deletes the entire cache, but
+//! it is not expected for this scenario to happen too often, and the only
+//! consequence is that one side or the other encounters an error and needs to
+//! retry.
+//!
+//! [^1]: A minor caveat is that downloads will delete an existing `src`
+//!   directory if it was extracted via an old cargo. See
+//!   [`crate::sources::registry::RegistrySource::unpack_package`]. This
+//!   should probably be fixed, but is unlikely to be a problem if the user is
+//!   only using versions of cargo with the same deletion logic.
+
+use super::FileLock;
+use crate::CargoResult;
+use crate::Config;
+use anyhow::Context;
+use std::cell::RefCell;
+use std::io;
+
+/// The style of lock to acquire.
+#[derive(Copy, Clone, Debug)]
+pub enum CacheLockMode {
```

Comment:
Whether on the enum or the variants, I feel like this could be a little more caller-focused in helping callers know what lock type they should choose.

Reply:
I'm not quite understanding what isn't clear or what could be added. If the caller is downloading packages to the cache, then it should have the `DownloadExclusive` lock. If the caller is reading from the cache (like when a build is running), then it should have the `Shared` lock. If the caller is modifying the cache, it should have `MutateExclusive`, as it mentions an example as garbage collection. I feel like each variant is saying how to choose it, so I'm not understanding what is missing.

Reply:
When I looked over the PR this was pulled out of, I was confused for a while as to why `cargo add` and `cargo remove` were `DownloadExclusive`.  I'm not downloading anything to the cache (though `cargo add` downloads to the index).  `cargo remove` only downloads in corner cases.  Looking over this PR, its clearer to me now but when someone comes p and writes code, they're likely to either pick an over-restrictive mode or just play error whack-a-mole until they get it right.

Reply:
Each variant now gives guidance on when to use it.

Reply:
Thanks!